### ISSUE TITLE
docker-php-ext-install fails with phpize not finding config.m4

### ISSUE THREAD

**Issue by k0pernikus** [docker-library/php#233]
(I wrote a question on [stackoverflow](http://stackoverflow.com/questions/37527803/how-to-install-extension-for-php-via-docker-php-ext-install/37527960#37527960) about it, but I guess it fits more here.)

My Dockerfile:

```
FROM php:5.6-apache

RUN php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');" && \
    php -r "if (hash_file('SHA384', 'composer-setup.php') === '070854512ef404f16bac87071a6db9fd9721da1684cd4589b1196c3faf71b9a2682e2311b36a5079825e155ac7ce150d') { echo 'Installer verified'; } else { echo 'Installer corrupt'; unlink('composer-setup.php'); } echo PHP_EOL;" && \
    php composer-setup.php && \
    php -r "unlink('composer-setup.php');"


RUN apt-get update && \
    apt-get install vim git -y
RUN docker-php-ext-install mysqlnd pdo pdo_mysql

RUN  cd /  && \
    git clone --depth=1 git://github.com/phalcon/cphalcon.git && \
    cd cphalcon/build && \
    ./install

RUN echo "extension=phalcon.so" > /usr/local/etc/php/conf.d/phalcon.ini
RUN a2enmod rewrite
```

My docker-compose.yml:

```
phlaconapp:
    hostname: phaclonapp
    dockerfile: Dockerfile
    build: ./
    ports:
        - "1080:80"
        - "1043:433"
    environment:
        TERM: xterm-color
        ENVIRONMENT: dev
    volumes:
        - ./:/var/www/html/
    links:
        - mysql
mysql:
    image: mysql:5.6
    volumes:
        - ./docker/mysql.d:/etc/mysql/conf.d
    ports: ["3306:3306"]
    environment:
        MYSQL_ROOT_PASSWORD: 'root'
```

Yet when running `docker-compose build` I get:

```
\mysql uses an image, skipping
Building phlaconapp
Step 1 : FROM php:5.6-apache
 ---> 2dd2b6f11998
Step 2 : RUN php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');" &&     php -r "if (hash_file('SHA384', 'composer-setup.php') === '070854512ef404f16bac87071a6db9fd9721da1684cd4589b1196c3faf71b9a2682e2311b36a5079825e155ac7ce150d') { echo 'Installer verified'; } else { echo 'Installer corrupt'; unlink('composer-setup.php'); } echo PHP_EOL;" &&     php composer-setup.php &&     php -r "unlink('composer-setup.php');"
 ---> Using cache
 ---> be0059709c8b
Step 3 : RUN apt-get update &&     apt-get install vim git -y
 ---> Using cache
 ---> 64cc739633c9
Step 4 : RUN docker-php-ext-install mysqlnd pdo pdo_mysql
 ---> Running in 8587aeb52c5b
+ cd /usr/src/php/ext/mysqlnd
+ phpize
Cannot find config.m4. 
Make sure that you run '/usr/local/bin/phpize' in the top level source directory of the module

ERROR: Service 'phlaconapp' failed to build: The command '/bin/sh -c docker-php-ext-install mysqlnd pdo pdo_mysql' returned a non-zero code: 1
```

How am I supposed to install extension with `docker-php-ext-install`?

**Comment by michaellopez** [comment]
We were hit by this today too. We've tracked it down to https://github.com/docker-library/php/commit/e3db777ec694977683c69d641ec2dcdc3bf41ce3.

Basically `docker-php-ext-install` now cleans up dependencies before exiting. Meaning that you can't chain a `docker-php-ext-configure` anymore, as `docker-php-ext-install` is not handling the dependencies. I believe this is a bug. `docker-php-ext-configure` should install the dependencies too.

Or how is this suppose to work now @tianon @ncopa?

**Comment by michaellopez** [comment]
I now see that you aren't using any configure step, but I think its related at least :)

**Comment by michaellopez** [comment]
My current workaround is not looking very nice: 

```
RUN ...
    && apk add --no-cache --virtual .phpize-deps $PHPIZE_DEPS \
    && docker-php-ext-configure gd --with-png-dir=/usr/include --with-freetype-dir=/usr/include/ --with-jpeg-dir=/usr/include/ \
    && apk del .phpize-deps \
```

**Comment by michaellopez** [comment]
I also just noticed that this workaround is needed for `pecl` installs too. What's the plan for handling that?

**Comment by michaellopez** [comment]
Our problem will get fixed by #239

**Comment by shouze** [comment]
yup so @k0pernikus maybe you can close this issue? 😉

**Comment by ivanjaros** [comment]
Is this still an issue? I have build the same image before but suddenly I just got:
```
Build complete.
Don't forget to run 'make test'.

Installing shared extensions:     /usr/local/lib/php/extensions/no-debug-non-zts-20160303/
Installing header files:           /usr/local/include/php/
find . -name \*.gcno -o -name \*.gcda | xargs rm -f
find . -name \*.lo -o -name \*.o | xargs rm -f
find . -name \*.la -o -name \*.a | xargs rm -f
find . -name \*.so | xargs rm -f
find . -name .libs -a -type d|xargs rm -rf
rm -f libphp.la       modules/* libs/*
+ cd date
+ phpize
Cannot find config.m4.
Make sure that you run '/usr/local/bin/phpize' in the top level source directory of the module
```

I tried installing everything separately but that takes too long since each step deletes and install required libs.

**Comment by estahn** [comment]
Seems to be a bug in `phpize` (https://bugs.php.net/bug.php?id=53571).

I added the following to `docker-php-ext-configure`:

```
/usr/local/bin # diff -u docker-php-ext-configure.bak docker-php-ext-configure
--- docker-php-ext-configure.bak
+++ docker-php-ext-configure
@@ -54,5 +54,6 @@

 set -x
 cd "$ext"
+[[ ! -f "config.m4" && -f "config0.m4" ]] && mv config0.m4 config.m4
 phpize
 ./configure "$@"
```

**Comment by allofmex** [comment]
Another Quick-And-Dirty solution for "phpize, Cannot find config.m4.":

#First we try to install your extension, it will fail here so force exit code 0 to keep Dockerfile processing.
#We do this to have the extension files downloaded for step 2
RUN docker-php-ext-install zlib; exit 0
#Now we rename the in step 1 downloaded file to desired filename
RUN cp  /usr/src/php/ext/zlib/config0.m4  /usr/src/php/ext/zlib/config.m4
#And try to install extension again, this time it works
RUN docker-php-ext-install zlib

**Comment by HazCod** [comment]
Any ideas when this will be fixed? Still having this issue.

**Comment by falmar** [comment]
For me it failed couple times trying to install other extensions like zlib openssl, but these were already included in the base image when I ran phpinfo. on `php:7.1.7-fpm-alpine`

Dockerfile
```dockerfile
FROM php:7.1.7-fpm-alpine
WORKDIR /usr/share/nginx/html
RUN apk --no-cache add --virtual .build-deps $PHPIZE_DEPS \
  && apk --no-cache add --virtual .ext-deps libmcrypt-dev freetype-dev \
  libjpeg-turbo-dev libpng-dev libxml2-dev msmtp postgresql-dev \
  && docker-php-source extract \
  && docker-php-ext-configure gd --with-freetype-dir=/usr/include/ \
                                   --with-png-dir=/usr/include/ \
                                   --with-jpeg-dir=/usr/include/ \
  && docker-php-ext-install gd mcrypt mysqli pdo pdo_mysql pdo_pgsql pgsql zip ftp opcache \
  && pecl install mongodb redis xdebug \
  && docker-php-ext-enable mongodb \
  && docker-php-ext-enable redis \
  && docker-php-ext-enable xdebug \
  && docker-php-source delete \
  && apk del .build-deps \
  # composer taken from (https://github.com/geshan/docker-php-composer-alpine)
  && apk --no-cache add curl git openssh \
  && curl -sSL https://getcomposer.org/installer | php -- --install-dir=/usr/bin --filename=composer
```

**Comment by acejam** [comment]
I came across this on `php:7.1.10-apache` while trying to install the `zlib` extension. The fix that @allofmex suggested worked for me.

**Comment by chrBrd** [comment]
Just ran into this issue trying to install OpenSSL on `php:7-fpm-alpine`. I'm not running `docker-php-ext-configure`.

**Comment by yosifkit** [comment]
@chrBrd, openssl and a few other modules are built with php and do not require `docker-php-ext*` scripts:  https://github.com/docker-library/php/blob/d8d798be28c49ced47d772aa899c0a1febabdba3/7.1/alpine3.4/fpm/Dockerfile#L116-L132

You can even verify that it is available:
```console
$ docker run --rm php:7-fpm-alpine php -i | grep -i ssl
Configure Command =>  './configure'  '--build=x86_64-linux-musl' '--with-config-file-path=/usr/local/etc/php' '--with-config-file-scan-dir=/usr/local/etc/php/conf.d' '--disable-cgi' '--enable-ftp' '--enable-mbstring' '--enable-mysqlnd' '--with-curl' '--with-libedit' '--with-openssl' '--with-zlib' '--with-pcre-regex=/usr' '--enable-fpm' '--with-fpm-user=www-data' '--with-fpm-group=www-data' 'build_alias=x86_64-linux-musl'
Registered Stream Socket Transports => tcp, udp, unix, udg, ssl, sslv3, tls, tlsv1.0, tlsv1.1, tlsv1.2
SSL => Yes
SSL Version => OpenSSL/1.0.2k
core SSL => supported
extended SSL => supported
openssl
OpenSSL support => enabled
OpenSSL Library Version => OpenSSL 1.0.2k  26 Jan 2017
OpenSSL Header Version => OpenSSL 1.0.2k  26 Jan 2017
Openssl default config => /etc/ssl/openssl.cnf
openssl.cafile => no value => no value
openssl.capath => no value => no value
Native OpenSSL support => enabled
```

**Comment by chrBrd** [comment]
Thanks for letting me know @yosifikit. Sorry I didn't notice that myself, especially since I'm sitting here with the Dockerfile open in another tab.

**Comment by tianon** [comment]
Closing, since https://github.com/docker-library/php/pull/239 solved the original issue described in this report. :+1:

**Comment by o10g** [comment]
Still doesn't work. 
sqlite3 has config0.m4 so phpize doesn't manage to build the extension. 
How it supposed to work?

**Comment by tianon** [comment]
@o10g see https://github.com/docker-library/php/issues/574:

```console
$ docker pull php:7
7: Pulling from library/php
Digest: sha256:560fa3ea47759beb28ed786eafdbc2d0e7091fc4179480dab72241b4972c7c4e
Status: Image is up to date for php:7

$ docker run -it --rm php:7 bash
root@8193d9dd2e8b:/# php -i | grep -i sqlite
PDO drivers => sqlite
pdo_sqlite
PDO Driver for SQLite 3.x => enabled
SQLite Library => 3.20.1
sqlite3
SQLite3 support => enabled
SQLite3 module version => 7.2.2
SQLite Library => 3.20.1
sqlite3.extension_dir => no value => no value
```

**Comment by andreaswarnaar** [comment]
Removing the docker-php-ext-install zlib from the docker fixed for me

**Comment by Lusitaniae** [comment]
Still affecting `php:7.4-fpm-alpine` but skipping installation of extensions that are already included in the image seems to avoid the issue.

**Comment by MatMercer** [comment]
Removing `docker-php-ext-install zlib` fix the issue with `php:7.3-fpm-alpine`.

If you [take a look here](https://github.com/docker-library/php/blob/74175669f4162058e1fb0d2b0cf342e35f9c0804/7.3/alpine3.13/fpm/Dockerfile#L146), the extension is already installed.

**Comment by shakaran** [comment]
In my case, I have this same problem having a Ubuntu Linux system, but using a ntfs partion for the docker data. With NTFS has failures in deleting files, so when I create a ext4 partition for store the docker data was working perfectly.

Why doesn't NTFS work for docker-data?
Docker needs to use OverlayFS (as seen in the overlay mounts on ...), which requires a native Linux file system (such as ext4, xfs, btrfs, etc.).

The ntfs3 driver:

Doesn't properly support extended attributes (xattr), security keys (security.*), and advanced Linux kernel operations.

Although you can technically mount NTFS with ntfs3, Docker and OverlayFS don't work properly on top of NTFS.
